### ISSUE TITLE
Optimise the main-less asm/IR case

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust-playpen#124]
A non-trivial number of my uses of play are to check the ASM/IR of a function (and I've seen others using it for that too). Currently the best way to do it is to add a `#![crate_type = "lib"]` at the top of the code which is kinda annoying. We could try to detect when the code is likely to be used for this purpose (e.g. no `main` or `#[test]`), and, say, compile in `lib` mode, gray out `[Run]` and even switch ctrl-enter to `[asm]`.

Thoughts?

**Comment by nodakai** [comment]
Let's begin with adding a checkbox for the "lib" mode before trying something smartish. [Run] can be coerced to [asm] when it's on.

**Comment by jonas-schievink** [comment]
Yes, I'd prefer not to do any magic detection, but rather have an additional group like "Crate Type" next to "Mode" and "Channel" where you can select between "bin" and "lib" or something like that (which would then replace or disable the Run button - personally I think graying out the Run button would be the best idea).

**Comment by jonas-schievink** [comment]
Okay, there's already a heuristic for `#[test]` vs `fn main()` (TIL the playpen can run tests). It doesn't seem like a good idea to gray out Run if neither was found (or does it? It certainly prevents running valid Rust code), but is certainly a possibility.

This also might not work too well with the "bin"/"lib" selection buttons, but the toolbar is almost completely full on my machine anyways :P

**Comment by nodakai** [comment]
Hold on... were there anything wrong with unconditionally adding `--crate-type=lib` to [ASM], [LLVM IR] and [MIR]?

**Comment by jonas-schievink** [comment]
@nodakai No, I think that could be a great start!
